rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java

```diff
@@ -211,23 +208,30 @@ private boolean allowByMinerNotBanned(Peer sender, Message message) {
      * record event if message is repeated
      */
     private boolean allowByMessageUniqueness(Peer sender, Message message) {
+        if (message.getMessageType() != MessageType.BLOCK_MESSAGE && message.getMessageType() != MessageType.TRANSACTIONS) {
+            return false;
+        }
+
         Keccak256 encodedMessage = new Keccak256(HashUtil.keccak256(message.getEncoded()));
 
-        boolean contains = receivedMessages.contains(encodedMessage);
+        if (!this.receivedPeerMessages.containsKey(sender)) {
+            this.receivedPeerMessages.put(sender, Collections.synchronizedSet(new HashSet<>()));
+        }
 
-        if (!contains) {
-            if (message.getMessageType() == MessageType.BLOCK_MESSAGE || message.getMessageType() == MessageType.TRANSACTIONS) {
-                if (this.receivedMessages.size() >= MAX_NUMBER_OF_MESSAGES_CACHED) {
-                    this.receivedMessages.clear();
-                }
-                this.receivedMessages.add(encodedMessage);
-            }
+        Set<Keccak256> receivedMessages = receivedPeerMessages.get(sender);
```

Comment:
You may want to use `Map.computeIfAbsent()` instead, it is even recommended by Sonar.